feat: Implement series remote data source with TMDB API

skaik-mo opened the pull request:
This commit introduces the `SeriesRemoteDataSource` and its implementation `TmdbSeriesApiRemoteDataSource` for fetching series data from The Movie Database (TMDB) API.

Key changes:
- Added `SeriesRemoteDataSource` interface defining methods for fetching series by name, genre, and retrieving genres.
- Implemented `TmdbSeriesApiRemoteDataSource` using Ktor for HTTP requests.
- Created `RequestBuilder` to encapsulate Ktor request logic and handle responses/errors.
- Defined `BaseRequest` for constructing API requests.
- Added DTOs for series (`SeriesDto`, `SeriesInfoDto`) and genres (`GenreDto`, `GenresResponse`).
- Included various `NetworkRequestException` types for error handling.
- Configured Ktor client and serialization in the `series` module's `build.gradle.kts`.
- Integrated `SeriesRemoteDataSource` into the DI graph in `dataModule.kt`.
- Added necessary Ktor and Kotlinx Serialization dependencies.

skaik-mo committed:
feat: Implement series remote data source with TMDB API

This commit introduces the `SeriesRemoteDataSource` and its implementation `TmdbSeriesApiRemoteDataSource` for fetching series data from The Movie Database (TMDB) API.

Key changes:
- Added `SeriesRemoteDataSource` interface defining methods for fetching series by name, genre, and retrieving genres.
- Implemented `TmdbSeriesApiRemoteDataSource` using Ktor for HTTP requests.
- Created `RequestBuilder` to encapsulate Ktor request logic and handle responses/errors.
- Defined `BaseRequest` for constructing API requests.
- Added DTOs for series (`SeriesDto`, `SeriesInfoDto`) and genres (`GenreDto`, `GenresResponse`).
- Included various `NetworkRequestException` types for error handling.
- Configured Ktor client and serialization in the `series` module's `build.gradle.kts`.
- Integrated `SeriesRemoteDataSource` into the DI graph in `dataModule.kt`.
- Added necessary Ktor and Kotlinx Serialization dependencies.

skaik-mo committed:
Remove gradle.properties from Git tracking

skaik-mo committed:
Refactor: Enhance series remote data source and DI setup

This commit refactors the `series` remote data source by:

- Simplifying the `RequestBuilder` to directly return the response body or throw an exception, removing the `ApiResult` wrapper.
- Making `RequestBuilder`, `BaseRequest` and `TmdbSeriesApiRemoteDataSource` internal.
- Removing unused exception classes from `NetworkRequestException.kt`.
- Introducing `SeriesResponse` and `GenresResponse` data classes.
- Updating `BaseRequest` to include default authorization and accept headers and retrieve the API key from `BuildConfig`.
- Creating `seriesDataModule` for Koin dependency injection, providing necessary dependencies for the series remote data source.
- Integrating `seriesDataModule` into the main application's Koin setup.
- Adding `gradle.properties` to `.gitignore`.
- Updating the Kotlin serialization plugin version.
- Adding `TMDB_API_KEY` to `buildConfigField` in `series/build.gradle.kts`.
- Removing series-related Koin definitions from the main `dataModule.kt` as they are now handled by `seriesDataModule`.

Mostafa-khaled21 commented on datasource/remote/series/src/main/java/com/giraffe/series/api/BaseRequest.kt:
```diff
@@ -0,0 +1,56 @@
+    var endPoint: String = ""
+    var method: HttpMethod = HttpMethod.Companion.Get
+    var headers: Map<String, String> = mapOf(
+        "Authorization" to "Bearer ${BuildConfig.TMDB_API_KEY}",
```
BuildConfig.TMDB_API_KEY is unresolved

Mostafa-khaled21 commented on datasource/remote/series/src/main/java/com/giraffe/series/api/BaseRequest.kt:
```diff
@@ -0,0 +1,56 @@
+
+    fun method(method: HttpMethod) = apply { this.method = method }
+
+    fun parameters(params: Map<String, Any>) = apply { this.parameters = params }
```
These functions (addHeader, parameters, body) appear to be unused in the current PR

Mostafa-khaled21 commented on datasource/remote/series/src/main/java/com/giraffe/series/model_dto/SeriesDto.kt:
```diff
@@ -0,0 +1,36 @@
+
+@Serializable
+data class SeriesDto(
+    @SerialName("id")
```
Remove unnecessary @SerialName annotations

skaik-mo committed:
Merge branch 'develop' into feature/series-remote

* develop:
  increase use cases test coverage
  test all explore usecases
  move safe calling to the repo and handle rquest codes in the remote datasource
  move safe calling to the repo and handle rquest codes in the remote datasource
  Update test_coverage.yml
  fix gradle secrets
  fix secrets file name
  update gradle/ app to use secrets from secrets.properties
  move secrets to the app module
  accept the page as parameter for remote request
  update the usecases to use execute instead invoke operator function
  add secrets to github action
  Remove secret.properties from tracking
  remove test code from main activity
  implement explore feature initial use cases
  refactor di modules to be separated
  implement explore repository
  add remote explore datasource
  setup explore feature local search history cache
  add explore feature modules

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/java/com/giraffe/cineverseapp/CineVerseApp.kt
#	app/src/main/java/com/giraffe/cineverseapp/di/databaseModule.kt

skaik-mo committed:
Merge branch 'feature/series' into feature/series-remote

* feature/series:
  feat: add fake series function and implement tests for series use cases
  feat: change releaseYear type to String in Season and Series data classes, and update SeriesException class hierarchy
  feat: update series repository to include store functionality and refactor use cases for better operator invocation
  feat: update series repository interface to return lists instead of results and add custom exception classes
  feat: add use cases for searching series by genre and name, and update SeriesRepository interface
  feat: implement use cases for fetching series by genre and name, and add Season and SeriesGenre data classes

skaik-mo committed:
Refactor: Relocate series data source and models

This commit relocates the `SeriesRemoteDataSource` interface and the `SeriesDto` and `GenreDto` data classes from the `datasource/remote/series` module to the `repository/series` module.

Additionally, the following changes were made:

- Updated `TmdbSeriesApiRemoteDataSource` to use the new location of `SeriesRemoteDataSource` and the model classes.
- Modified `TmdbSeriesApiRemoteDataSource` to accept `genreId` as `Int` instead of `String`.
- Updated `BaseRequest` to accept `baseURL` and `accessToken` in its constructor and removed unused methods.
- Updated `build.gradle.kts` files to reflect the new module dependencies and plugin usage.
- Renamed Koin module from `seriesDataModule` to `seriesRemoteDataModule`.
- Adjusted Koin module to retrieve `BASE_URL` and `ACCESS_TOKEN` using named qualifiers.
- Removed explicit `@SerialName` annotations from DTOs where the property name matches the JSON key.
- Reordered fields in `SeriesDto`.

skaik-mo committed:
Merge branch 'feature/series' into feature/series-remote

* feature/series:

skaik-mo committed:
Merge branch 'refs/heads/feature/series' into feature/series-remote

# Conflicts:
#	gradle/libs.versions.toml
#	repository/series/build.gradle.kts
#	repository/series/src/main/java/com/giraffe/series/model/SeriesDto.kt

skaik-mo committed:
Merge branch ' feature/series' into feature/series-remote

skaik-mo committed:
Merge branch 'feature/series' into feature/series-remote

* feature/series:
  fix: update search cache table name and queries for consistency

# Conflicts:
#	repository/series/src/main/java/com/giraffe/series/model/CachedSearchCacheDto.kt
#	repository/series/src/main/java/com/giraffe/series/util/DatabaseConstants.kt
